server/mailet/dkim/src/main/java/org/apache/james/jdkim/mailets/MimeMessageHeaders.java

```diff
@@ -20,50 +20,61 @@
 package org.apache.james.jdkim.mailets;
 
 import java.util.Enumeration;
-import java.util.HashMap;
-import java.util.LinkedList;
 import java.util.List;
-import java.util.Map;
+import java.util.Locale;
 
+import javax.mail.Header;
 import javax.mail.MessagingException;
 import javax.mail.internet.MimeMessage;
 
+import com.github.fge.lambdas.Throwing;
+import org.apache.commons.lang3.tuple.Pair;
 import org.apache.james.jdkim.api.Headers;
 
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Streams;
+
 /**
  * An adapter to let DKIMSigner read headers from MimeMessage
  */
 final class MimeMessageHeaders implements Headers {
 
-    private final Map<String, List<String>> headers;
+    private final ImmutableListMultimap<String, String> headers;
     private final List<String> fields;
 
     public MimeMessageHeaders(MimeMessage message) throws MessagingException {
-        headers = new HashMap<>();
-        fields = new LinkedList<>();
-        for (Enumeration<String> e = message.getAllHeaderLines(); e.hasMoreElements();) {
-            String head = e.nextElement();
-            int p = head.indexOf(':');
-            if (p <= 0) {
-                throw new MessagingException("Bad header line: " + head);
-            }
-            String headerName = head.substring(0, p).trim();
-            String headerNameLC = headerName.toLowerCase();
-            fields.add(headerName);
-            List<String> strings = headers.get(headerNameLC);
-            if (strings == null) {
-                strings = new LinkedList<>();
-                headers.put(headerNameLC, strings);
-            }
-            strings.add(head);
-        }
+        ImmutableList<Pair<String, String>> headsAndLines = Streams.stream(Iterators.forEnumeration(message.getAllHeaderLines()))
```

Comment:
do we have a good testsuite covering all these refactorings?

Reply:
current one was good enough